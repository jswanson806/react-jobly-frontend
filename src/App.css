body {
  font-family: 'Raleway', sans-serif;
}

main {
  width: 100vw;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-top: 15vh;
  height: calc(100vh - 56px); /* router view - navbar height */
  overflow: scroll;
  color: #000000; /* black */

  /* apply colors to background */
  background-color: #c5a880; /* Khaki */
  background-image: linear-gradient(135deg, #c5a880 0%, #e6d8af 100%);
}

.card {
  border-radius: 0;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14),
    0 3px 1px -2px rgba(0, 0, 0, 0.12), 0 1px 5px 0 rgba(0, 0, 0, 0.2);
  margin-bottom: 2em;

  /* apply color to border */
  border: 2px solid #8c4646; /* Firebrick */
}
