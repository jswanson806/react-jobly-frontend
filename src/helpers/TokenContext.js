import React, { createContext, useState } from 'react';

// Create a new context for the token
export const TokenContext = createContext();

// Define the initial state for the token
const INITIAL_TOKEN_STATE = '';

// Create a provider component for the token context
export const TokenProvider = ({ children }) => {
  // Set up state to store the token value
  const [token, setToken] = useState(INITIAL_TOKEN_STATE);

  // Render the provider component and pass the token value and setter to the context
  return (
    <TokenContext.Provider value={{ token, setToken }}>
      {children}
    </TokenContext.Provider>
  );
};
