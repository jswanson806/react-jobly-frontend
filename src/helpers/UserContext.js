// Import the necessary modules from React
import React, { createContext, useState } from 'react';

// Define the initial state for the user
const INITIAL_USER_STATE = {};

// Create a new context for the user
export const UserContext = createContext();

// Create a provider component for the user context
export const UserProvider = ({ children }) => {
  // Set up state to store the user object
  const [user, setUser] = useState(INITIAL_USER_STATE);

  // Render the provider component and pass the user object and setter to the context
  return (
    <UserContext.Provider value={{ user, setUser }}>
      {children}
    </UserContext.Provider>
  );
};

