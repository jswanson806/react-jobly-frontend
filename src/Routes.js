import React, { useState, useEffect, useContext } from 'react';
import { Route, Switch, Redirect } from 'react-router-dom';
import Home from './Home.js';
import CompanyList from './CompanyList.js';
import JobsList from './JobsList.js'
import CompanyDetails from './CompanyDetails.js';
import LogInForm from './LogInForm.js';
import SignupForm from './SignupForm.js';
import UpdateProfileForm from './UpdateProfileForm.js';
import NotFound from './NotFound.js';
import JoblyApi from './api.js';
import FormContext from './helpers/FormContext.js';
import { TokenContext } from './helpers/TokenContext.js';

/**
 * Routes Component
 *
 * Renders the main routing logic for the application, mapping different routes to their respective components.
 * It also manages the state of jobs and companies fetched from the database, as well as handling form filtering.
 * Requires the `TokenContext` and `FormContext` to be provided by a parent component.
 */


const Routes = () => {
    const {token} = useContext(TokenContext);

    const loggedIn = !token || token === "" ? false : true
    //
    const [jobs, setJobs] = useState([]);
    // tracks state of companies
    const [ companies, setCompanies ] = useState([]);

    // gets companies and from the database and adds them to companies state
    useEffect(() => {
        async function getCompanies() {
            let res = await JoblyApi.getCompanies();
            setCompanies(res.companies);
        };
        // calls the getCompanies async function
        getCompanies();
    }, [])

    // gets all jobs from the database and adds them to jobs state
    useEffect(() => {
        async function getJobs() {
            const res = await JoblyApi.getJobs();
            setJobs(res.jobs)
        };
        // calls the getjobs async function
        getJobs();
    }, [])

    const filterItems = async (filters) => {
        let res = '';
        // if filters includes 'title' it is a filter for jobs
        Object.keys(filters).includes('title') 
        ? 
        res = await JoblyApi.getJobs(filters) 
        // otherwise, it is a filter for companies
        : 
        res = await JoblyApi.getCompanies(filters)

        return res;
    }

    // checks the state of the page loading and renders a loading message if true
    if(jobs.length === 0 || companies.length === 0) {
         return <p data-testid="Loading">Loading &hellip;</p>;
    };

    return (
            <FormContext.Provider value={{filterItems}} data-testid="formcontext">
                <main data-testid="routes">
                    <Switch>
                        <Route exact path="/">
                            <Home />
                        </Route>
                        <Route exact path="/companies">
                            {loggedIn ? <CompanyList companies={companies} filterItems={filterItems}/>  : <Redirect to="/login" />}
                        </Route>
                        <Route exact path="/companies/:handle">
                            {loggedIn ? <CompanyDetails /> : <Redirect to="/login" />}
                        </Route>
                        <Route exact path="/jobs">
                            {loggedIn ? <JobsList jobs={jobs} filterItems={filterItems} /> : <Redirect to="/login" />}
                        </Route>
                        <Route exact path="/login">
                            <LogInForm />
                        </Route>
                        <Route exact path="/signup">
                            <SignupForm />
                        </Route>
                        <Route exact path="/profile">
                            {loggedIn ? <UpdateProfileForm />
                            : <Redirect to="/login" />}
                        </Route>
                        <Route>
                            <NotFound />
                        </Route>
                    </Switch>
                </main>
            </FormContext.Provider>
    )
}

export default Routes;