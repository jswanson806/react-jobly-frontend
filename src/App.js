import React, { useContext, useEffect } from 'react';
import { useHistory } from "react-router-dom";
import './App.css';
import Routes from './Routes.js';
import NavBar from './NavBar.js';
import JoblyApi from './api.js';
import AppContext from './helpers/AppContext.js';
import { UserContext } from './helpers/UserContext.js';
import { TokenContext } from './helpers/TokenContext.js';

/**
 * App Component
 *
 * The root component of the application. Manages user authentication, routing, and global state.
 *
*/

function App() {

  const history = useHistory();

  // ********************* State *********************

  const {user, setUser} = useContext(UserContext);
  const {token, setToken} = useContext(TokenContext);

  // ********************* Hooks *********************

  // keep token saved between page refreshes
  // set token to the value in localStorage
  useEffect(() => {
    // get the value of 'token' from localStorage
    const storedToken = localStorage.getItem('token');
    const storedUser = localStorage.getItem('user');

    // check for value of stored token to be different than stateful token
    if(storedToken !== token && storedToken !== null) {
      // set token to the value in localStorage
      setToken(storedToken);
    }
  // check for value of stored user to be different than stateful user
    if(storedUser !== user && storedUser !== null) {
      // turn the stringified user object in localStorage into an object and set user state
      setUser(JSON.parse(storedUser))
    }

    // only run on first render
  }, [])

  // ********************* Functions *********************

  // registers a new user and logs them in
  const registerUser = async (userInfo) => {
    // api method registers user
    await JoblyApi.registerUser(userInfo);
    // calls logIn function to log in the newly registered user
    logIn(userInfo);
  }

  // logs in an existing user
  const logIn = async ({username, password}) => {
    // authenticates user
    let resToken = await JoblyApi.logInUser(username, password);
    // set token in localStorage and state
    localStorage.setItem('token', resToken);
    setToken(resToken);
    // setLocalToken(resToken)
    // sets api token to token received after authentication
    JoblyApi.token = resToken;
    // get the user from the database
    let userRes = await JoblyApi.getUser(username);
    // stringify user object and set user info in localStorage and state
    localStorage.setItem('user', JSON.stringify(userRes.user));
    setUser(userRes.user);
  }

  // logs out user
  const logOut = async () => {
    // reset api token
    JoblyApi.token = "";
    // clear localStorage'token' and 'user'
    localStorage.removeItem('token');
    localStorage.removeItem('user');
    // reset state of 'token' and 'user'
    setToken("")
    setUser("")
    // redirect to home
    history.push('/');
  }

  // calls api to patch user info
  const updateUserProfile = async (newUserInfo) => {
    const {applications} = user;
    await JoblyApi.patchUser(newUserInfo);
    // set user info state and user info in localStorage to new user info
    // res.user.applications = user.applications
    newUserInfo.applications = applications;
    setUser(newUserInfo);
  }

  // ********************* Return *********************

  return (
    // AppContext provides functions and user and token state to child components
    <AppContext.Provider value={{user, setUser, token, setToken, registerUser, logIn, logOut, updateUserProfile}} >
          <div data-testid="app">
            <NavBar />
            <Routes />
          </div>
    </AppContext.Provider>
  )
}

export default App;