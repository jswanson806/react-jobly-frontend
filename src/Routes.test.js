import React from 'react';
import { render, act, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Routes from './Routes.js'
import { TokenContext } from './helpers/TokenContext.js';
import JoblyApi from './api.js';
import { MemoryRouter } from 'react-router-dom';
import App from './App.js';
import { UserContext } from './helpers/UserContext.js';
import AppContext from './helpers/AppContext.js';

let token = "testToken";

let user = {
    username: "testuser",
    password: "password",
    firstName: "firstName",
    lastName: "lastName",
    email: "test@email.com",
    applications: []
  }

jest.mock('./api.js')



beforeEach(() => {
    JoblyApi.getCompanies.mockResolvedValueOnce({ companies: [""] });
    JoblyApi.getJobs.mockResolvedValueOnce({ jobs: [""] });
})
/** ********************* smoke / snapshot tests ********************* */

describe('routes', () => {

    it("should render without crashing", async () => {

        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={token}>
                    <MemoryRouter>
                        <Routes />
                     </MemoryRouter>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })
    });

    it("should match snapshot", async () => {

        let asFragment;

        await act(async () => {
            const {asFragment: fragment} = render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <MemoryRouter>
                        <Routes />
                     </MemoryRouter>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
            asFragment = fragment;
        })
        expect(asFragment()).toMatchSnapshot();
    })
})


describe("Routes correctly", () => {

    it("should route to '/'", async () => {
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <MemoryRouter initialEntries={['/']}>
                        <Routes />
                     </MemoryRouter>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Home-content")).toBeInTheDocument();
    })

    it("should route to '/companies'", async () => {
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <MemoryRouter initialEntries={['/companies']}>
                        <Routes />
                     </MemoryRouter>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Company-list")).toBeInTheDocument();
    })

    it("should route to '/companies/:handle'", async () => {
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <MemoryRouter initialEntries={['/companies/testcompany']}>
                        <Routes />
                     </MemoryRouter>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Company-detail")).toBeInTheDocument();
    })

    it("should route to '/jobs'", async () => {
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <MemoryRouter initialEntries={['/jobs']}>
                        <Routes />
                     </MemoryRouter>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Job-list")).toBeInTheDocument();
    })

    it("should route to '/login'", async () => {
        const logIn = jest.fn();
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <AppContext.Provider value={{logIn}}>
                        <MemoryRouter initialEntries={['/login']}>
                            <Routes />
                        </MemoryRouter>
                    </AppContext.Provider>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Login-form")).toBeInTheDocument();
    })

    it("should route to '/signup'", async () => {
        const registerUser = jest.fn();
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <AppContext.Provider value={{registerUser}}>
                        <MemoryRouter initialEntries={['/signup']}>
                            <Routes />
                        </MemoryRouter>
                    </AppContext.Provider>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Signup-form")).toBeInTheDocument();
    })

    it("should route to '/profile'", async () => {
        const registerUser = jest.fn();
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <AppContext.Provider value={{registerUser}}>
                        <MemoryRouter initialEntries={['/profile']}>
                            <Routes />
                        </MemoryRouter>
                    </AppContext.Provider>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Profile-form")).toBeInTheDocument();
    })

    it("should render NotFound if no url matches", async () => {
        const registerUser = jest.fn();
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <AppContext.Provider value={{registerUser}}>
                        <MemoryRouter initialEntries={['/notValidUrl']}>
                            <Routes />
                        </MemoryRouter>
                    </AppContext.Provider>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Not-found")).toBeInTheDocument();
    })

})

describe("Routes redirects", () => {

    it("should redirect from '/companies' to '/login' if user is logged out", async () => {
        const logIn = jest.fn();
        token = "";
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <AppContext.Provider value={{logIn}}>
                        <MemoryRouter initialEntries={['/companies']}>
                            <Routes />
                        </MemoryRouter>
                    </AppContext.Provider>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Login-form")).toBeInTheDocument();

    })

    it("should redirect from '/jobs' to '/login' if user is logged out", async () => {
        const logIn = jest.fn();
        token = "";
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <AppContext.Provider value={{logIn}}>
                        <MemoryRouter initialEntries={['/jobs']}>
                            <Routes />
                        </MemoryRouter>
                    </AppContext.Provider>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Login-form")).toBeInTheDocument();

    })

    it("should redirect from '/companies/:handle' to '/login' if user is logged out", async () => {
        const logIn = jest.fn();
        token = "";
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <AppContext.Provider value={{logIn}}>
                        <MemoryRouter initialEntries={['/companies/testcompany']}>
                            <Routes />
                        </MemoryRouter>
                    </AppContext.Provider>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Login-form")).toBeInTheDocument();

    })

    it("should redirect from '/profile' to '/login' if user is logged out", async () => {
        const logIn = jest.fn();
        token = "";
        await act(async () => {
            render(
                <UserContext.Provider value={{user}}>
                <TokenContext.Provider value={{token}}>
                    <AppContext.Provider value={{logIn}}>
                        <MemoryRouter initialEntries={['/profile']}>
                            <Routes />
                        </MemoryRouter>
                    </AppContext.Provider>
                </TokenContext.Provider>
                </UserContext.Provider>
            );
        })

        expect(screen.getByTestId("Login-form")).toBeInTheDocument();

    })
    
})