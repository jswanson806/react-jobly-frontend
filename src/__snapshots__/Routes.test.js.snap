// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes should match snapshot 1`] = `
<DocumentFragment>
  <main
    data-testid="routes"
  >
    <div
      class="Home-content"
      data-testid="Home-content"
    >
      <h1
        class="Home-title"
      >
        Jobly
      </h1>
      <p
        class="Home-text"
      >
        All the jobs in one, convenient place.
      </p>
      <h2
        class="Home-greeting"
        data-testid="Home-greeting-message"
      >
        Welcome Back, testuser!
      </h2>
    </div>
  </main>
</DocumentFragment>
`;
